### ISSUE TITLE
CMake script errors

### ISSUE THREAD

**Issue by syb0rg** [erikd/libsndfile#136]
Upon running `cmake ..` in an out of source build directory, I encounter the following errors:

```
CMake Error at CMake/build.cmake:28 (add_executable):
  Cannot find source file:

    src/test_endswap.c

  Tried extensions .c .C .c++ .cc .cpp .cxx .m .M .mm .h .hh .h++ .hm .hpp
  .hxx .in .txx
Call Stack (most recent call first):
  CMakeLists.txt:237 (lsf_build_src_test_c)


CMake Error at CMake/build.cmake:44 (add_executable):
  Cannot find source file:

    tests/utils.c

  Tried extensions .c .C .c++ .cc .cpp .cxx .m .M .mm .h .hh .h++ .hm .hpp
  .hxx .in .txx
Call Stack (most recent call first):
  CMakeLists.txt:250 (lsf_build_test_c)


CMake Error at CMake/build.cmake:44 (add_executable):
  Cannot find source file:

    tests/floating_point_test.c

  Tried extensions .c .C .c++ .cc .cpp .cxx .m .M .mm .h .hh .h++ .hm .hpp
  .hxx .in .txx
Call Stack (most recent call first):
  CMakeLists.txt:255 (lsf_build_test_c)


CMake Error at CMake/build.cmake:44 (add_executable):
  Cannot find source file:

    tests/benchmark.c

  Tried extensions .c .C .c++ .cc .cpp .cxx .m .M .mm .h .hh .h++ .hm .hpp
  .hxx .in .txx
Call Stack (most recent call first):
  CMakeLists.txt:243 (lsf_build_test_c)


CMake Error at CMake/build.cmake:44 (add_executable):
  Cannot find source file:

    tests/pipe_test.c

  Tried extensions .c .C .c++ .cc .cpp .cxx .m .M .mm .h .hh .h++ .hm .hpp
  .hxx .in .txx
Call Stack (most recent call first):
  CMakeLists.txt:267 (lsf_build_test_c)


CMake Error at CMake/build.cmake:44 (add_executable):
  Cannot find source file:

    tests/header_test.c

  Tried extensions .c .C .c++ .cc .cpp .cxx .m .M .mm .h .hh .h++ .hm .hpp
  .hxx .in .txx
Call Stack (most recent call first):
  CMakeLists.txt:258 (lsf_build_test_c)


CMake Error at CMake/build.cmake:44 (add_executable):
  Cannot find source file:

    tests/pcm_test.c

  Tried extensions .c .C .c++ .cc .cpp .cxx .m .M .mm .h .hh .h++ .hm .hpp
  .hxx .in .txx
Call Stack (most recent call first):
  CMakeLists.txt:265 (lsf_build_test_c)


CMake Error at CMake/build.cmake:44 (add_executable):
  Cannot find source file:

    tests/scale_clip_test.c

  Tried extensions .c .C .c++ .cc .cpp .cxx .m .M .mm .h .hh .h++ .hm .hpp
  .hxx .in .txx
Call Stack (most recent call first):
  CMakeLists.txt:270 (lsf_build_test_c)


CMake Error at CMake/build.cmake:44 (add_executable):
  Cannot find source file:

    tests/rdwr_test.c

  Tried extensions .c .C .c++ .cc .cpp .cxx .m .M .mm .h .hh .h++ .hm .hpp
  .hxx .in .txx
Call Stack (most recent call first):
  CMakeLists.txt:269 (lsf_build_test_c)


CMake Error at CMake/build.cmake:44 (add_executable):
  Cannot find source file:

    tests/write_read_test.c

  Tried extensions .c .C .c++ .cc .cpp .cxx .m .M .mm .h .hh .h++ .hm .hpp
  .hxx .in .txx
Call Stack (most recent call first):
  CMakeLists.txt:280 (lsf_build_test_c)


CMake Error: CMake can not determine linker language for target: test_main
CMake Error: Cannot determine link language for target "test_main".
CMake Error: CMake can not determine linker language for target: dither_test
CMake Error: Cannot determine link language for target "dither_test".
CMake Error: CMake can not determine linker language for target: largefile_test
CMake Error: Cannot determine link language for target "largefile_test".
CMake Error: CMake can not determine linker language for target: floating_point_test
CMake Error: Cannot determine link language for target "floating_point_test".
CMake Error: CMake can not determine linker language for target: cpp_test
CMake Error: Cannot determine link language for target "cpp_test".
CMake Error: CMake can not determine linker language for target: chunk_test
CMake Error: Cannot determine link language for target "chunk_test".
CMake Error: CMake can not determine linker language for target: channel_test
CMake Error: Cannot determine link language for target "channel_test".
CMake Error: CMake can not determine linker language for target: ulaw_test
CMake Error: Cannot determine link language for target "ulaw_test".
CMake Error: CMake can not determine linker language for target: ogg_test
CMake Error: Cannot determine link language for target "ogg_test".
CMake Error: CMake can not determine linker language for target: aiff_rw_test
CMake Error: Cannot determine link language for target "aiff_rw_test".
CMake Error: CMake can not determine linker language for target: fix_this
CMake Error: Cannot determine link language for target "fix_this".
CMake Error: CMake can not determine linker language for target: alaw_test
CMake Error: Cannot determine link language for target "alaw_test".
CMake Error: CMake can not determine linker language for target: benchmark
CMake Error: Cannot determine link language for target "benchmark".
CMake Error: CMake can not determine linker language for target: checksum_test
CMake Error: Cannot determine link language for target "checksum_test".
CMake Error: CMake can not determine linker language for target: command_test
CMake Error: Cannot determine link language for target "command_test".
CMake Error: CMake can not determine linker language for target: external_libs_test
CMake Error: Cannot determine link language for target "external_libs_test".
CMake Error: CMake can not determine linker language for target: compression_size_test
CMake Error: Cannot determine link language for target "compression_size_test".
CMake Error: CMake can not determine linker language for target: dwvw_test
CMake Error: Cannot determine link language for target "dwvw_test".
CMake Error: CMake can not determine linker language for target: error_test
CMake Error: Cannot determine link language for target "error_test".
CMake Error: CMake can not determine linker language for target: pipe_test
CMake Error: Cannot determine link language for target "pipe_test".
CMake Error: CMake can not determine linker language for target: format_check_test
CMake Error: Cannot determine link language for target "format_check_test".
CMake Error: CMake can not determine linker language for target: headerless_test
CMake Error: Cannot determine link language for target "headerless_test".
CMake Error: CMake can not determine linker language for target: header_test
CMake Error: Cannot determine link language for target "header_test".
CMake Error: CMake can not determine linker language for target: locale_test
CMake Error: Cannot determine link language for target "locale_test".
CMake Error: CMake can not determine linker language for target: lossy_comp_test
CMake Error: Cannot determine link language for target "lossy_comp_test".
CMake Error: CMake can not determine linker language for target: multi_file_test
CMake Error: Cannot determine link language for target "multi_file_test".
CMake Error: CMake can not determine linker language for target: misc_test
CMake Error: Cannot determine link language for target "misc_test".
CMake Error: CMake can not determine linker language for target: pcm_test
CMake Error: Cannot determine link language for target "pcm_test".
CMake Error: CMake can not determine linker language for target: scale_clip_test
CMake Error: Cannot determine link language for target "scale_clip_test".
CMake Error: CMake can not determine linker language for target: peak_chunk_test
CMake Error: Cannot determine link language for target "peak_chunk_test".
CMake Error: CMake can not determine linker language for target: raw_test
CMake Error: Cannot determine link language for target "raw_test".
CMake Error: CMake can not determine linker language for target: stdio_test
CMake Error: Cannot determine link language for target "stdio_test".
CMake Error: CMake can not determine linker language for target: rdwr_test
CMake Error: Cannot determine link language for target "rdwr_test".
CMake Error: CMake can not determine linker language for target: sfversion
CMake Error: Cannot determine link language for target "sfversion".
CMake Error: CMake can not determine linker language for target: stdin_test
CMake Error: Cannot determine link language for target "stdin_test".
CMake Error: CMake can not determine linker language for target: win32_ordinal_test
CMake Error: Cannot determine link language for target "win32_ordinal_test".
CMake Error: CMake can not determine linker language for target: stdout_test
CMake Error: Cannot determine link language for target "stdout_test".
CMake Error: CMake can not determine linker language for target: string_test
CMake Error: Cannot determine link language for target "string_test".
CMake Error: CMake can not determine linker language for target: virtual_io_test
CMake Error: Cannot determine link language for target "virtual_io_test".
CMake Error: CMake can not determine linker language for target: win32_test
CMake Error: Cannot determine link language for target "win32_test".
CMake Error: CMake can not determine linker language for target: write_read_test
CMake Error: Cannot determine link language for target "write_read_test".

```

This behavior occurs despite running `./autogen.sh cmake` before this command, and has been replicated on Windows, Mac, and Linux machines.

**Comment by erikd** [comment]
Does it work in the source directory?

**Comment by syb0rg** [comment]
@erikd Nope, I tried that as well (across all platforms) and got the same results.  In place of the first file it throws an error on (`src/test_endswap.c`), there is `src/test_endswap.def` and `src/test_endswap.tpl`.

**Comment by erikd** [comment]
Its been a while since I looked at the CMake stuff. It was always experimental, it never got top a place where I was happy with it. 

The probelm you are running into is that I use a tool called `autogen` for templating  C test suite code and that never got hooked into the CMake based build system.

**Comment by erikd** [comment]
Ok, I've put a couple of hours into improving the CMake build system. Specifically commit 9c6e45a10304f6ece90e888674b82c49c82770b9.

Please test.

**Comment by syb0rg** [comment]
Sorry for the slow response, I'm a lot less active on weekends.  Testing on my Mac, I get the following error:

```
CMake Error at CMake/file.cmake:32 (if):
  if given arguments:

    "LESS" "1"

  Unknown arguments specified
Call Stack (most recent call first):
  CMakeLists.txt:39 (assert_line_count_non_zero)

```

I am unable to test on the other systems right now.

**Comment by erikd** [comment]
Like I said, its only likely to work on Linux. Patches welcome.

**Comment by syb0rg** [comment]
I'm working on overhauling your entire CMake system, but it's gonna take me a little bit (especially for testing).  I'm trying to remove your current CMake dependencies on Autotools, as that is what is inhibiting your portability right now.

**Comment by erikd** [comment]
@syb0rg My idea is that to generate `CMake/config.h.in` you need Autotools, but that once that is done, the autotools should not be needed. I am very much in favour of Don't Repeat Yourself.

**Comment by syb0rg** [comment]
Understood, but that isn't the full reason for the rewrite,  I'm refactoring down the code a bit as well.

**Comment by erikd** [comment]
I would strongly advise small steps to begin with.

**Comment by syb0rg** [comment]
I've gotten everything to compile nicely, and produce a library.  There are no tests or executables generated yet.  Since CMake is experimental for you, should I submit a pull request now or should I add these things first?

**Comment by erikd** [comment]
First of all, the Stack Overflow post is someone's *opinion* and that opinion is almost the opposite of *my* opinion.

The existing Autotool based build system has been working extremely well for over a decade. I am 100% certain that there are subtleties that you are not even aware of. In fact, the *only* problem I am aware of is that it cannot be used with Microsoft Visual Studio. Since I don't use MSVS, I see that as a small problem.

The problems with the CMake build system currently in my github tree are as follows:

* Requires autotools to bootstrap.
* Doesn't build on Mac OS X.
* Doesn't build on Windows.
* Doesn't allow cross compiling.
* Doesn't hide non-public internal library symbols.
* Doesn't detect the availablity of GCC warning flags and add the ones that are supported.
* Doesn't allow GCC/Clang stack protector be be enabled/disabled.
* Doesn't detect CPU clipping modes.
 
and thats just the things I know about. There's probably others.

CMake will not become the primary build system for libsndfile until it is both feature competitive with the the Autotools based system as well as being easier to maintain and extend.

I would however like to see what you have. I suggest you put it up one github and let me pull it down to have a look.

**Comment by erikd** [comment]
I've found your branch. Having a look at it now.

**Comment by syb0rg** [comment]
Some good points, I'd like to comment on them to give you my perspective:

- "I am 100% certain that there are subtleties that you are not even aware of." - With Autotools or with CMake? Probably true for both anyhoo.

- "Requires autotools to bootstrap" - for now, yes.  My rewrite is attempting to drop this.  I know you are a fan of DRY principals (as am I), but there is bound to be a bit of repetition if we plan to support two build systems going forward as their objective is to do the same thing.

- "Doesn't build on Mac OS X." - I was able to build it on my Mac, albeit hard-coding a few options to make sure I had everything working at first.  I plan on making this more versatile in the future.

- "Doesn't build on Windows." - I actually just built the `.dll` on my Windows machine while writing this response!  But like with my Mac, that was with the hard-coded options.

-  "Doesn't allow cross compiling." - Cross compiling is supported by CMake starting with version 2.6.0.  I can implement this for you in the future if you would like it.

- "Doesn't hide non-public internal library symbols." - I'm confused a bit what you mean by this.  Perhaps because my CMake build doesn't have them?

- "Doesn't detect the availablity of GCC warning flags and add the ones that are supported." - I've done this in some other GitHub repo's, [such as this one](https://github.com/syb0rg/Khronos/blob/master/CMakeLists.txt#L32).  I do agree that compiler handling could be implemented better by CMake.

- "Doesn't allow GCC/Clang stack protector be be enabled/disabled." - You could write a conditional setting `-fstack-protector-all` or `-fno-stack-protector`.

- "Doesn't detect CPU clipping modes."  - I'm not really sure what CPU clipping is, and am struggling to find anything on it with Google.  Could you clarify this a bit more?

- "CMake will not become the primary build system for libsndfile until it is both feature competitive with the the Autotools based system as well as being easier to maintain and extend." - It shouldn't be long :wink:

**Comment by syb0rg** [comment]
Since the initial issue here has been "fixed" and you have #71 open for adding CMake, I'm going to close this thread and comment over there as needed.
